[CodeGen] Really renumber slot indexes before register allocation

jayfoad opened the pull request:
PR #66334 tried to renumber slot indexes before register allocation, but
the numbering was still affected by list entries for instructions which
had been erased. Fix this to make the register allocator's live range
length heuristics even less dependent on the history of how instructions
have been added to and removed from SlotIndexes's maps.

jayfoad committed:
[CodeGen] Really renumber slot indexes before register allocation

PR #66334 tried to renumber slot indexes before register allocation, but
the numbering was still affected by list entries for instructions which
had been erased. Fix this to make the register allocator's live range
length heuristics even less dependent on the history of how instructions
have been added to and removed from SlotIndexes's maps.

jayfoad commented:
TODO: update 33 failing tests with manual checks.

asl commented:
MSP430 changes seems to be fine

jayfoad commented on llvm/lib/CodeGen/SlotIndexes.cpp:
```diff
@@ -238,8 +238,27 @@ void SlotIndexes::repairIndexesInRange(MachineBasicBlock *MBB,
+        // been erased. We have to update these entries but we don't want them
+        // to affect the rest of the slot numbering, so set them to half way
+        // between the neighboring real instrucion indexes.
+        I->setIndex(Index + SlotIndex::InstrDist / 2);
```
Fair point. I do need to understand this better. I'm not even sure about the "LiveIntervals may still refer to entries for instructions that have been erased" comment - I wrote this when I was still in the middle of debugging some failures, so I may have misunderstood what was going on.

jayfoad replied:
I have looked into this some more. There is a related to this comment in SlotIndexes::removeMachineInstrFromMaps/removeSingleMachineInstrFromMaps:
```
// FIXME: Eventually we want to actually delete these indexes.
```
Machine verification would fail if any live interval endpoint referred to a deleted instruction so I think this is just a transient problem. There are various bits of code that temporarily use the slot index for a deleted instruction. Here's one simple example where `OldIndex` is used (in an assertion) after `MI` has been deleted: https://github.com/llvm/llvm-project/blob/10edd5d9436153ace82009a04900ac67d3adc202/llvm/lib/CodeGen/LiveIntervals.cpp#L1507

`RegisterCoalescer::joinVirtRegs` seems to have more complicated problems when it calls into `LiveRange::join`: 
https://github.com/llvm/llvm-project/blob/10edd5d9436153ace82009a04900ac67d3adc202/llvm/lib/CodeGen/RegisterCoalescer.cpp#L3685
At this point, `LHS` seems to refer to dead instructions as a result of the call to `LHSVals.pruneValues` above, which will only be fixed when we call `LiveIntervals::extendToIndices` below.

qcolombet replied:
Maybe add a comment that `SlotIndex` pointing to empty "instruction" should not be visible from the outside world.
In other words, to avoid weird behavior, users of this API shouldn't call it when old references of `Slotindex`s are kept around, e.g., getting a `SlotIndex` for an instruction, remove that instruction from the map while still using this `SlotIndex` to do thing.

jayfoad replied:
Ah, I understand a bit more now.

First there are transient problems where live ranges refer to deleted instructions (I guess this why [the "poison" machinery](https://github.com/llvm/llvm-project/commit/4cf8da94198dcb3af2e8eefa1559bfe6c0c332dc) was never used). But we can document that packIndexes should not be called when there are transient problems.

Second, insertMachineInstrInMaps will pick a new index half between two adjacent entries in indexList, even if one of these entries is for a deleted instruction. That is why we have to pick sane index values for those entries here. But a better alternative might be to remove them from indexList here (and add them to a free list for later reuse?).

jayfoad replied:
Hi @qcolombet, the latest version of this patch removes entries from indexList that refer to deleted instructions. Are you happy with that? As far as I can see these entries are never used for anything. They should really have been removed in removeMachineInstrFromMaps but (as the FIXME comment there says) we can't do that because of the "transient brokenness" problem. packIndexes seems like a good place to clean them up when we know there are no transient problems.

blackgeorge-boom replied:
@jayfoad This new functionality works great for me. I see now that you also skip the first slot in the MBB range, which is a block slot, and you also handle the last slot of the MBB range separately, so that you don't accidentally remove those. I don't know a lot about potential transient problems to comment further.

jayfoad replied:
> I see now that you also skip the first slot in the MBB range, which is a block slot

Yes because it the same entry as the last slot for the previous MBB range. The first slot for the first MBB in the function should always have index 0 so I don't bother to renumber it - I guess I could add an assert for that to make it more obvious.

blackgeorge-boom replied:
> The first slot for the first MBB in the function should always have index 0 so I don't bother to renumber it - I guess I could add an assert for that to make it more obvious.

That would be great actually!

jayfoad replied:
> > The first slot for the first MBB in the function should always have index 0 so I don't bother to renumber it - I guess I could add an assert for that to make it more obvious.
> 
> That would be great actually!

Done.

qcolombet reviewed (approved):
LGTM

jayfoad commented:
Rebased and fixed almost all lit tests. (Sorry for the force push but quite a few lit tests conflicted with other upstream changes since this PR was opened.)

I might need help with the last two lit test failures:
```
Failed Tests (2):
  LLVM :: CodeGen/Hexagon/reg-scavengebug-2.ll
  LLVM :: CodeGen/Hexagon/swp-epilog-phi7.ll
```

jayfoad commented:
> I might need help with the last two lit test failures:
> 
> ```
> Failed Tests (2):
>   LLVM :: CodeGen/Hexagon/reg-scavengebug-2.ll
>   LLVM :: CodeGen/Hexagon/swp-epilog-phi7.ll
> ```

Hi @kparzysz-quic, could you help with these please? The code changes quite a bit, so I'm not even sure if they are still testing what they were supposed to test.

nikic commented:
@jayfoad I'd suggest marking those tests as XFAIL, to avoid further bitrot.

jayfoad commented:
> @jayfoad I'd suggest marking those tests as XFAIL, to avoid further bitrot.

Done, thanks.

blackgeorge-boom commented on llvm/lib/CodeGen/SlotIndexes.cpp:
```diff
@@ -247,8 +253,8 @@ void SlotIndexes::packIndexes() {
       if (I.getInstr()) {
-        Index += SlotIndex::InstrDist;
         I.setIndex(Index);
+        Index += SlotIndex::InstrDist;
```
Also now the post-increment of `Index` is much more intuitive.

jayfoad commented:
Any more comments? If not I will plan to merge this on Monday, based on @qcolombet's approval.
